layer4/connection.go

```diff
@@ -83,7 +83,9 @@ func (cx *Connection) Read(p []byte) (n int, err error) {
 			cx.bufReader = nil
 			err = nil
 		}
-		return
+		if !(n == 0 && err == nil) { // prevent first read from returning 0 bytes with no err
+			return
+		}
```

Comment:
It is just unnecessary, instead we can easily advance to a real Read. It was annoying during debugging and also interfered with `TestConnection_RecordAndRewind`. So yes not really a bug, but it makes things easier.

Reply:
To be clear the 0 Read always happens at the first Read because how we setup the `bufReader` in `record`.

Reply:
Gotcha. That's fine then. I might suggest that the comment be moved to its own line and say something more specific:

```
// first read can be empty due to bufReader
```

or something like that?

Or, should we adjust `bufReader` to not return an empty read instead?

Reply:
I do not see an easy to understand way to init `bufReader` which prevents the 0 Read. The if condition before the return is easier to understand in my opinion, since its affect is more local.